### PR TITLE
keeping router GEMM in fp32 for deterministic inference (DeepSeek V3/V4)

### PR DESCRIPTION
**Pull request by xikronz**
wip for #34758 o-0

## Motivation

under deterministic inference the DeepSeek router GEMM falls back to `F.linear`, which stores buffers in bf16. Every other path in the `MoEGate.forward` already returns fp32 (`dsv3_router_gemm`, `linear_bf16_fp32`, prefill-CP). Those logits go into `noaux_tc` top-k over 256/384 experts with an fp32 correction bias, so using bf16's 8-bit mantissas here can change which experts are selected when scores sit within one ULP.

also, batch invariance cares about reduction order not output dtype. The deterministic matmul already accumulates in fp32 and has an fp32 store branch but it was unreachable because the output buffer was always allocated as the input dtype.

## Modifications

threads `out_dtype` through `matmul_persistent`, uses it from `_mm_dtype_compat` (instead of compute-then-`.to(out_dtype)`, which rounded fp32→bf16→fp32), and has the deterministic gate request fp32 for all token counts so it does not reintroduce the `max_router_gemm_tokens` kernel switch.

## Checklist

- [ ] Format your code according to the [Format code with pre-commit](https://docs.sglang.io/developer_guide/contribution_guide.html#format-code-with-pre-commit).
- [ ] Add unit tests according to the [Run and add unit tests](https://docs.sglang.io/developer_guide/contribution_guide.html#run-and-add-unit-tests).
- [ ] Update documentation according to [Write documentations](https://docs.sglang.io/developer_guide/contribution_guide.html#write-documentations).
- [ ] Provide accuracy and speed benchmark results according to [Test the accuracy](https://docs.sglang.io/developer_guide/contribution_guide.html#test-the-accuracy) and [Benchmark the speed](https://docs.sglang.io/developer_guide/contribution_guide.html#benchmark-the-speed).
- [ ] Follow the SGLang code style [guidance](https://docs.sglang.io/developer_guide/contribution_guide.html#code-style-guidance).

## Review and Merge Process

1. Ping Merge Oncalls to start the process. See the [PR Merge Process](https://github.com/sgl-project/sglang/blob/main/.github/MAINTAINER.md#pull-request-merge-process).
2. Get approvals from [CODEOWNERS](https://github.com/sgl-project/sglang/blob/main/.github/CODEOWNERS) and other reviewers.
3. Trigger CI tests with [comments](https://docs.sglang.io/developer_guide/contribution_guide.html#how-to-trigger-ci-tests) or contact authorized users to do so.
   - Common commands include `/tag-and-rerun-ci`, `/tag-run-ci-label`, `/rerun-failed-ci`
4. After green CI and required approvals, ask Merge Oncalls or people with Write permission to merge the PR.

---
### CI States

Latest PR Test (Base): :white_check_mark: [Run #34601801479](https://github.com/sgl-project/sglang/actions/runs/34601801479)
Latest PR Test (Extra): :x: [Run #34601801269](https://github.com/sgl-project/sglang/actions/runs/34601801269)
Latest PR Test (AMD ROCm 10): :x: [Run #34601801477](https://github.com/sgl-project/sglang/actions/runs/34601801477)

### CODE REVIEW

File: python/sglang/srt/models/deepseek_v2.py [hunk 1]
```diff
@@ -534,6 +510,12 @@ def forward(
             )
 
         if get_exec().deterministic.enable_deterministic_inference:
+            if _is_cuda or _is_hip:
+                from sglang.srt.batch_invariant_ops import matmul_persistent
+
+                return matmul_persistent(
```
Comment (b8zhong) [inline python/sglang/srt/models/deepseek_v2.py]: Is it possible to call:

```python
return torch.mm(hidden_states, self.weight.t(), out_dtype=torch.float32)
```

And it could be monkeypatched directly right?

File: python/sglang/srt/models/deepseek_v2.py [hunk 2]
```diff
@@ -534,6 +510,12 @@ def forward(
...
+                    hidden_states, self.weight.t(), out_dtype=torch.float32
+                )
             return F.linear(hidden_states, self.weight, None)
```
Comment (b8zhong) [inline python/sglang/srt/models/deepseek_v2.py]: Can we change it to FP32 (through casts)

File: python/sglang/srt/models/deepseek_v2.py (hunk 1 above)
Comment (xikronz) [reply python/sglang/srt/models/deepseek_v2.py]: ouuu good point

File: python/sglang/srt/models/deepseek_v2.py (hunk 2 above)
Comment (xikronz) [reply python/sglang/srt/models/deepseek_v2.py]: good catch

**Review by BBuf — APPROVED** [review]
